### PR TITLE
Add option to update an existing PR if it's already exist

### PR DESCRIPTION
**Pull request by genlu**
FYI @dibarbet @JoeRobich @RikkiGibson 

I was a little concerned about the case that it triggers a rebuild when existing run is almost finished, thus delay our codeflow. But looking at the time from PR creation to auto-merge in Roslyn, it usually happens in about 1.5 hours, well within the 3 hours window. So in most cases, we will only reset when CI failed or stuck (due to infra issue), which is actually desired.

Another change I'm considering is to rerun CI if it hasn't finished in certain time, which hopefully can mitigate some transient infra issue. Thoughts?

### CODE REVIEW

**Comment by genlu** [comment]
@JoeRobich Who else is using the merge tool? Probably need to let them know about this change before merge.

**Comment by JoeRobich** [comment]
from the [config ](https://github.com/dotnet/roslyn-tools/blob/master/src/GitHubCreateMergePRs/config.xml) it looks like project system, f#, test...

**Comment by genlu** [comment]
FYI @AbhitejJohn @mavasani @brettfo @jmarolf 
Do you have any concerns about this change?

**Comment by RikkiGibson** [comment]
If anyone is using the tool with CI runs that take 3+ hours they might be unhappy. If this is the case you could just make the tool check the `<repo>` element for an optional attribute that configures this behavior.

**Comment by brettfo** [comment]
How often does the tool run?  I'm concerned that I might be fixing merge conflicts/tests/etc. and the tool will re-run and overwrite my changes.  Or the tool could run a few minutes after I've pushed fixes and it'll reset the merge discarding my fixes.

**Comment by brettfo** [comment]
I'd prefer to keep the behavior as-is and for teams that want an updated merge, they can delete the stalled/bad/etc. merge and a new one will be created next run.

**Comment by JoeRobich** [comment]
Can we use the `merge conflicts` tag to skip updating PRs with conflicts?

**Comment by brettfo** [comment]
That still may not be sufficient; I might have test failures when merging feature branches.  This isn't uncommon in F#.

**Comment by genlu** [comment]
> How often does the tool run? 

Every 3 hours now

> make the tool check the <repo> element for an optional attribute

I like this idea, will add and enable only for Roslyn

> Can we use the merge conflicts tag to skip updating PRs with conflicts?

Like this too, will add.

**Comment by genlu** [comment]
I have pushed change as mentioned above. now we don't reset branch if:
1. There's "Merge Conflicts" label, or
2. "updateExistingPr" isn't set to true for the repo in the config file.

Could someone please explain how the "Merge Conflicts" label works? So once the conflict is resolved, is it supposed to be removed? By whom?

File: src/GithubMergeTool/GithubMergeTool.cs [hunk 1]
```diff
@@ -97,13 +93,57 @@ public class GithubMergeTool
...
+            var existingPR = JArray.Parse(await prsResponse.Content.ReadAsStringAsync()).FirstOrDefault(pr => (string)pr["title"] == prTitle);
+            if (existingPR != null)
+            {
+                if (updateExistingPr)
+                {
+                    // Get the SHA of the PR branch HEAD
+                    var prSha = ((JValue)existingPR["head"]["sha"]).ToObject<string>();
+                    var existingPrNumber = ((JValue)existingPR["number"]).ToObject<string>();
+
+                    if (prSha != srcSha)
```
Comment (RikkiGibson) [inline src/GithubMergeTool/GithubMergeTool.cs]: Does this assume that if the source branch is on a different SHA than the merge branch, there are new commits in the source branch? If you pushed a commit to resolve merge conflicts or test failures then would this code throw that away?

Perhaps if "force" were disabled then we would gracefully fail to update the merge PR in the case that new commits were added to it directly?

**Comment by RikkiGibson** [comment]
The Merge Conflicts label is pretty primitive. I implemented it in #624. At the time of PR creation, it just waits a few seconds for GitHub to compute the mergability status. Merge conflicts can still be introduced later on (e.g. from the base branch getting updated after the creation of the merge PR). The tool will not catch that.

Have you tried running the

**Comment by genlu** [comment]
> Have you tried running the

@RikkiGibson  Running what ? :)

File: src/GithubMergeTool/GithubMergeTool.cs (hunk 1 above)
Comment (genlu) [reply src/GithubMergeTool/GithubMergeTool.cs]: I'm not sure what we will end up with in the PR by using `reset --mixed`. We can do a `git merge --ff-only` instead, which would be a no-op if there's conflict resolving commit in the merge branch.

File: src/GithubMergeTool/GithubMergeTool.cs [hunk 2]
```diff
@@ -97,13 +93,54 @@ public class GithubMergeTool
...
+                        await PostComment(existingPrNumber, $"Reset HEAD of `{prBranchName}` to `{srcSha}`");
+
+                        // Check for merge conflicts again after reset.
+                        existingPrMergeable = await IsPrMergeable(existingPrNumber);
+                    }
+
+                    // Add label if there's merge conflicts even if we made no change to merge branch,
+                    // since can also be introduced by change in destination branch. It's no-op if the
+                    // label already exists. 
+                    if (existingPrMergeable == false)
```
Comment (genlu) [inline src/GithubMergeTool/GithubMergeTool.cs]: @RikkiGibson We are now checking mergeable state directly, and add "merge conflicts" label accordingly during update too.

**Comment by RikkiGibson** [comment]
> Running what ? :)

🤦‍♂ I have no idea what the end of that sentence was. Sorry..

File: src/GitHubCreateMergePRs/Program.cs [hunk 3]
```diff
@@ -13,6 +13,7 @@ public static async Task Main(string[] args)
         // Default when no args passed in.
         var isDryRun = true;
         var isAutomated = false;
+        var updateExistingPr = true;
```
Comment (RikkiGibson) [inline src/GitHubCreateMergePRs/Program.cs]: Should this default to 'false'?

Comment (RikkiGibson) [reply src/GitHubCreateMergePRs/Program.cs]: Actually, for simplicity it feels like this shouldn't be a command line arg, it should only be specified in config.xml per-repo. Do you agree?

Comment (genlu) [reply src/GitHubCreateMergePRs/Program.cs]: Sure, sounds reasonable.

**Comment by RikkiGibson** [comment]
Have you created a test PR using your version of the tool? One easy way to do this is to go in config.xml locally and change `<repo owner="dotnet" name="roslyn" ...` to your fork, and comment out other repos. With and without the new config setting to see if it behaves as expected (no need to wait 3 hours :smile: just run it consecutively with different configurations)

If there are no branches to merge in your fork you can also create dummy branches and have the config reference them.

**Comment by genlu** [comment]
@RikkiGibson I think I have addressed your prior comments, any more concerns?

**Pull request merged**